interface/views.py

```diff
@@ -350,14 +349,58 @@ def code_view(request, pk, filename):
     submission = get_object_or_404(Submission, pk=pk)
     all_tas = submission.assignment.course.teaching_assistants.all()
 
-    if request.user not in all_tas:
+    if request.user not in all_tas and request.user is not submission.user:
         return HttpResponse(status=403)
 
     tree = tree_view(request, submission)
 
+    if filename == "code":
+        context = {
+            "tree": tree,
+            "pk": submission.pk,
+            "sub": submission,
+            "path": filename,
+        }
+        return render(request, "interface/code_view_homepage.html", context)
+
     file = extract_file(request, submission, filename)
 
-    context = {"file_content": file.read(), "tree": tree, "pk": submission.pk}
+    context = {
+        "file_content": file.getvalue(),
+        "tree": tree,
+        "pk": submission.pk,
+        "sub": submission,
+        "path": filename,
+        "file_exists": False,
+    }
+
+    if (
+        file.getvalue() != "The file is missing!"
```

Comment:
Q: Could you save that value to don't having to call the same method twice?
like
```
file_value = file.getValue()
```
and then you can do
```
if file_value not in ["The file is missing!", "The archive is missing!"]:
```